### DISCUSSION TITLE
Finding changed elements

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#2445] [Q&A] [answered]
Hi. I'm trying to query for all the _Bis.Geometric3d_ elements which have been changed within a given change summary. However, I'm also trying to determine those elements which have had changes made to their associated element aspects. I'm attempting to do so with the query shown below, but performance-wise, it's not a great solution. It also takes far longer to do the _ChangedValueState.afterUpdate_ query than the _ChangedValueState.afterInsert_ query (hours vs minutes). Any ideas for better ways to query for _Bis.Geometric3d_ elements which have had either their direct properties changed, or their associated element aspects changed?

```
SELECT DISTINCT * FROM (
    SELECT
        a.ecInstanceId,
    FROM
        Bis.Geometric3d.Changes(${changeSummary.id}, ${changedValueState}) a

    UNION

    SELECT
        a.ecInstanceId,
    FROM
        Bis.ElementMultiAspect.Changes(${changeSummary.id}, ${changedValueState}) aspect
    JOIN
        Bis.Geometric3d a ON a.ecInstanceId = aspect.element.id

    UNION

    SELECT
        a.ecInstanceId,
    FROM
        Bis.ElementUniqueAspect.Changes(${changeSummary.id}, ${changedValueState}) aspect
    JOIN
        Bis.Geometric3d a ON a.ecInstanceId = aspect.element.id
)
```

**Comment by diegopinate** [comment, marked as answer]
Using the *.Changes is very slow as you have found. It is recommended to use `ecchange` queries instead if you only care for their instance Ids: https://www.itwinjs.org/learning/changesummaries/

Simple example:
```
SELECT ChangedInstance.Id, OpCode FROM ecchange.change.InstanceChange WHERE Summary.Id=?
```

In your case, here's how one of your SELECT queries would look like, which should be faster:

```
SELECT ChangedInstance.Id, OpCode FROM ecchange.change.InstanceChange WHERE Summary.Id=? And ChangedInstance.ClassId IS (bis.GeometricElement3d)
```

The "IS" keyword there is a "Type Filter Expression" that we don't have docs exposed for at the moment, I added those docs at the end of my reply if you are interested. This helps with polymorphism. Here's another example that would allow you to get elements that are both PUMP and VALVE class that have Bis.Element as a parent class:
```
SELECT * FROM bis.Element WHERE ECClassId IS (ONLY opm.PUMP, ONLY opm.VALVE)
```

Keep in mind that, if you want all properties, you will need to query for more than aspects. You may have other things in your schema that result in related property paths to the element. For that, you would need to use Presentation to obtain those paths. @grigasp could you comment on how that can be done in TS?

FWIW, we have plans to re-write some utilities we have that will make this much easier later on, but we don't have an ETA.

Here are the docs that we have not exposed yet related to type filter expressions:

# Type Filter Expression

ECSQL now allow to filter by type
```
    <ECClassId>|<expr> IS [NOT] ( [ONLY] <qualified-classname-1> [, [ONLY] <qualified-classname-2>, ...])
```
### Example
```sql
-- Returns row where element instance is only of type GeometricElement3d or GeometricElement2d
SELECT * FROM bis.Element WHERE ECClassId IS (bis.GeometricElement3d, bis.GeometricElement2d)

-- Returns only rows where instance is exactly of particular type
SELECT * FROM bis.Element WHERE ECClassId IS (ONLY opm.PUMP, ONLY opm.VALVE)

-- Invert the selection set
SELECT * FROM bis.Element WHERE ECClassId IS NOT (ONLY opm.PUMP)
```

**Reply by grigasp** [reply]
On `master` there's a [`PresentationManager.getContentSources`](https://github.com/imodeljs/imodeljs/blob/4bf6a8e20018f8f57c472e9a46f606facd52ba93/presentation/backend/src/presentation-backend/PresentationManager.ts#L523) function that can be used to get property sources for given classes (e.g. for `bis.GeometricElement3d`). Sadly there's no official release with this API yet, so you'll have to wait till `3.0` is released.

On `2.x` the only option is to use [`PresentationManager.getContentDescriptor`](https://github.com/imodeljs/imodeljs/blob/7f2a1667415e9399aafa1bc6c1f79872d3ff9184/presentation/backend/src/presentation-backend/PresentationManager.ts#L654) function with the following params:
```ts
rulesetOrId: "ElementProperties",
displayType: "",
keys: new KeySet([{ className: "bis.GeometricElement3d", id: Id64.invalid }]),
```
The resulting `Descriptor` object has a `selectClasses` attribute which matches what `getContentSources` would return. However, since building the content descriptor involves not only finding the classes, but also the properties, expect the performance to be much worse on large iModels.
